crypto/evp/p_lib.c

```diff
@@ -1188,6 +1188,26 @@ static int legacy_asn1_ctrl_to_param(EVP_PKEY *pkey, int op,
             }
             return rv;
         }
+    case ASN1_PKEY_CTRL_SUPPORTS_MD_NID:
+        {
+            const OSSL_PARAM *gettable = NULL;
+            OSSL_PARAM params[2];
+
+            if ((gettable = EVP_PKEY_gettable_params(pkey)) == NULL
+                    || OSSL_PARAM_locate_const(gettable,
+                                               OSSL_PKEY_PARAM_DIGEST_SUPPORTED) == NULL)
+                return -2;
+
+            /*
+             * We use EVP_PKEY_get_params to check an "assertion" that a certain
+             * digest is supported. We set the input OSSL_PARAM and expect the
+             * get_params will fail if that hash is not supported.
+             */
+            params[0] = OSSL_PARAM_construct_utf8_string(OSSL_PKEY_PARAM_DIGEST_SUPPORTED,
+                                                         (char *)OBJ_nid2sn(arg1), 0);
+            params[1] = OSSL_PARAM_construct_end();
+            return EVP_PKEY_get_params(pkey, params);
```

Comment:
Getting a parameter overwrites it's contents, in this case the short name.  I don't think this is desirable.

Reply:
The output OSSL_PARAMs are irrelevant. The code tests whether a given short name is in the list of supported digests, so only the int result of EVP_PKEY_get_params is used.

I agree this is weird, but with only GET and SET operations there is no easy way to determine if a certain value (the short name) is in a certain list on the provider side (the list of supported digests). This is what the CTRL operation was doing for the legacy and there seems to be no equivalent.

An alternate solution would be to retrieve all supported digests and check whether the one is in the list. This list may include various synonyms (SHA256, SHA-256, SHA2-256) so it will not be short. But I can do it that way, if you agree that would be better. It's probably a cleaner solution than misusing GET for doing an assertion.

BTW, what would be your preferred way to obtain a list? A set of OSSL_PARAMs with identical names, or e.g. a semicolon separated list such as `"SHA256:SHA-256:SHA2-256"`.

BTW2, it is not possible to simply fetch all digest algorithms provided by the PKEY's provider because the individual PKEY may be restricted only for usage with a specific digest. The list of supported digests is a property of a PKEY, not the provider itself.

Reply:
However, we are very strict about "get" parameters being received and "set" parameters being sent.  This is doing both with a "get" parameter.  This is **wrong**.